client/src/modules/authorization/components/SignUpForm.jsx

```diff
@@ -0,0 +1,309 @@
+    this.setState(
+      { password, isPasswordValid: isValid },
+      this.comparePasswords
+    );
+  };
+
+  onPasswordConfirmChange = (passwordConfirm, isValid) => {
+    this.setState(
+      { passwordConfirm, isPasswordConfirmValid: isValid },
+      this.comparePasswords
+    );
+  };
+
+  nameValidator = value => {
+    const { isEmpty, isLength } = validator;
+
+    if (isEmpty(value)) {
+      return 'authorization.input.username.empty';
+    }
+
+    if (!isLength(value, { min: 1, max: 32 })) {
+      return 'authorization.input.username.invalid';
+    }
+
+    return '';
+  };
+
+  emailValidator = value => {
+    const { isEmpty, isEmail } = validator;
+
+    if (isEmpty(value) || !isEmail(value)) {
+      return 'authorization.input.email.invalid';
+    }
+
+    return '';
+  };
+
+  passwordValidator = value => {
+    const { matches } = validator;
+
+    if (!matches(value, /^[^\s]{4,32}$/)) {
+      return 'authorization.input.password.invalid';
+    }
+
+    return '';
+  };
+
+  comparePasswords = () => {
+    const {
+      higherLevelErrors,
+      higherLevelErrors: { passwordConfirmError },
+      password,
+      passwordConfirm
+    } = this.state;
+    let newError;
+
+    if (passwordConfirmError && password === passwordConfirm) {
+      newError = '';
+    }
+
+    if (
+      !passwordConfirmError &&
+      passwordConfirm !== undefined &&
+      password !== passwordConfirm
+    ) {
+      newError = 'authorization.input.password.not-match';
+    }
+
+    if (newError !== undefined) {
+      this.setState({
+        higherLevelErrors: {
+          ...higherLevelErrors,
+          passwordConfirmError: newError
+        }
+      });
+    }
+  };
+
+  isFormValid = () => {
+    const {
+      higherLevelErrors,
+      isEmailValid,
+      isNameValid,
+      isPasswordConfirmValid,
+      isPasswordValid
+    } = this.state;
+
+    const isError = _find(higherLevelErrors, error => error !== '');
+
+    if (
+      isNameValid &&
+      isEmailValid &&
+      isPasswordValid &&
+      isPasswordConfirmValid &&
+      !isError
+    ) {
+      return this.setState({ isFormValid: true });
+    }
+
+    this.setState({ isFormValid: false });
+  };
+
+  handleSignUp = () => {
+    const { signUp } = this.props;
+    const { email, name, password, passwordConfirm } = this.state;
+
+    this.setState({ pending: true });
+
+    this.pendingPromise = makeAbortablePromise(
+      signUp(email, name, password, passwordConfirm)
+    );
+    this.pendingPromise.promise
+      .then(() => this.setState({ confirmationSend: true, pending: false }))
+      .catch(err => {
+        if (!(err instanceof AbortPromiseException)) {
+          const { status } = err.response;
+          const newState = { pending: false };
+
+          if (status === 406) {
+            const { errors } = err.response;
+            newState.higherLevelErrors = errors;
+          }
+
+          this.setState(newState);
```

Comment:
why not just this.setState({pending: false})??

Reply:
because of if backend validation - if backend returns some errors I have to save them in the state. So I created the newState object and add pending to that object and if errors exist add those errors to it. In the end, I save this object as a state once.